reference/replica-operations.md

```diff
@@ -0,0 +1,74 @@
+
+### Scenario Four
+
+A cluster has three Mayastor Nodes \(MSN\) deployed; "Node-1", "Node-2" and "Node-3".  Each MSN hosts two MayastorPools \(MSP\) and currently, no MayastorVolumes \(MSV\) have been defined.  MSN Node-1 hosts MSPs "Pool-1-A" and "Pool-1-B", whilst MSN Node-2 hosts "Pool-2-A and "Pool-2-B" and Node-3 hosts "Pool-3-A" and "Pool-3-B".  A single MSV exists in the cluster, which has a replica count of 3.  The volume's replicas are all healthy and are located on Pool-1-A, Pool-2-A and Pool-3-A.  An application is using the volume, so all replicas are receiving I/O.
+
+The physical disk of Pool-3-A suffers a permanent failure, causing all failure of all I/O sent to the replica it hosts.
+
+Expected Behaviour:  The volume will enter and remain in the `degraded` state.  The replica in Pool-3-A will be in the state `CHILD_FAULTED`, as observed in the volume's MSV custom resource.  No replica replacement \(nor subsequent rebuild\) will occur, since Rule 3 states that the faulted replica hosted in Pool-3-A on Node-3 remains a part of that volume's nexus and therefore the other pool on Node-3 \(Pool-3-B)\ cannot be selected as a location for its replacement because of Rule 2.
+
+### Scenario Five
+
+Given the post-disk failure situation of Scenario four, the user edits the MSV custom resource for the volume, reducing the value of `Replica Count` from 3 to 2
+
+Expected Behaviour:  The control plane will reconcile the actual \(replicas=3\) vs desired \(replicas=2\) state of the volume by removing a replica from its nexus.  The replica located on Node-3 will be the one selected, since it is in the `CHILD_FAULTED` state whilst the other two replicas are healthy \(`CHILD_ONLINE`)\.  This is Rule 5 in action.  Once the faulted replica has been retired, the volume state will become `healthy` again.
+
+### Scenario Six
+
+In scenario Five, after editing the MSV the user waits for the volume state to become `healthy` again.  The desired and actual replica count are now 2.  The volume's replicas are located in MSPs on both Node-1 and Node-2. The user then edits the MSV custom resource again, increasing  the `Replica Count:` from 2 to 3 again.
+
+Expected Behaviour:  The volume's state will become `degraded`, reflecting the difference in desired vs actual replica count.  The control plane will select Pool-3-B on Node-3 as the location for the new replica required.  It is following Rule 2; the replica previously in Pool-3-A, which failed, has already been retired from the membership of the volume's nexus as a result of the user's actions in Scenario Five.  Node-3 is therefore again a suitable candidate and has an available, online pool with sufficient capacity \(Pool-3-B which was unaffected by the previous disk failure in Pool-3-A\), which will be selected as the location of this replica.  Once the replica has been created, a rebuild will take place and eventually, the volume state will return to `healthy` but now with a `Replica Count` of 3.
```

Comment:
Good point, this opens up a situation I'd not considered.  Seeing as scenario six flows from scenario 3 where "The physical disk of Pool-3-A suffers a permanent failure", we'd want pool-3-b be chosen.  However my own testing suggests that, at least from the perspective of the MSP resource, this failure is 'not known' to the control plane.  It is therefore possible that the failed pool would be selected?  If the replica creation fails as I expect it would be in this situation (the pool metadata cannot be written to), would moac attempt to select a new candidate pool, or would the replica count increment just fail at this point?  Or would k8s continue to attempt the reconciliation?  If a replica creation in a pool fails, will the MSP record the pool as faulted?  So if the reconciliation loop continues, it would converge on a successful replacement in pool-3-b?

Reply:
I think these are all really valid questions. Unfortunately I'm not certain of the behaviour in MOAC with regards to pools. I think we should test these cases to confirm the behaviour - especially with regards to the interactions with k8s (crds, retries etc.).

Reply:
I agree.  I'll schedule some additional tests targeting this scenario before making any updates to this doc.